cupy/core/include/cupy/float16.cuh

```diff
@@ -0,0 +1,182 @@
+#pragma once
+#ifdef CUPY_CUB_BLOCK_REDUCTION
+#include <cupy/complex.cuh>
+#endif
+
+
+// math
+#ifndef M_PI
+#define M_PI 3.1415926535897932384626433832795
+#endif
+
+#ifdef __HIPCC__
+
+#include <hip/hip_fp16.h>
+
+#elif __CUDACC_VER_MAJOR__ >= 9
+
+#include <cuda_fp16.h>
+
+#else  // #if __CUDACC_VER_MAJOR__ >= 9
```

Comment:
note: This part will be able to remove after https://github.com/cupy/cupy/issues/3301 is resolved.

Reply:
Yeah would be great. We have this macro check in several places.